File: Algorithm Toolbox/Programming Challenges/ProgrammingChallenges/ProgrammingChallenges/Week 2 Algorithmic Warmup/Fibonacci.swift
//
//  Fibonacci.swift
//  ProgrammingChallenges
//
//  Created by Santiago Pelaez Rua on 7/08/20.
//  Copyright © 2020 Santiago Pelaez Rua. All rights reserved.
//

import Foundation

class Fibonacci: CLIExecutable {
    static func execute() {
        print("Enter what nth fibonacci number to calculate (0 <= n <= 45)")
        let n = Int(readLine()!)!
        assert(0 <= n && n <= 45, "n is out of the limits")
        print(fibonacci(n: n))
    }
    
    static func fibonacci(n: Int) -> Int {
        if n <= 1 {
            return n
        }
        
        var numbers = [0, 1]
        
        for i in 2...n {
            numbers.append(numbers[i-1] + numbers[i-2])
        }
        
        return numbers[n]
    }
}

File: Algorithm Toolbox/Programming Challenges/C#/fibonacci.cs
using System;
namespace Fibonacci {
    public class Program {

        static void Main(string [] args) {
            var input = Console.ReadLine();
            var n = int.Parse(input);

            Console.Write(fibonacci(n));
        }

        static int fibonacci(int n) {
            if (n <= 1) {
                return n;
            }

            int[] numbers = new int[n + 1];
            numbers[0] = 0;
            numbers[1] = 1;

            for (int i = 2; i <= n; ++i) {
                numbers[i] = numbers[i - 1] + numbers[i - 2];
            }

            return numbers[n];
        }
    }
}
